File: src/main/java/net/maple/handlers/user/UserDropMesoRequestHandler.java
package net.maple.handlers.user;

import client.Character;
import client.Client;
import field.object.drop.MesoDrop;
import net.maple.handlers.PacketHandler;
import util.packet.PacketReader;

public class UserDropMesoRequestHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();
        reader.readInteger(); // ??
        int meso = reader.readInteger();

        if (meso <= chr.getMeso() && meso > 9 && meso < 50001) {
            chr.gainMeso(-meso);
            MesoDrop drop = new MesoDrop(chr.getId(), chr, meso, 0);
            drop.setPosition(chr.getPosition());
            chr.getField().enter(drop);
        }
    }
}

File: src/main/java/net/maple/handlers/user/UserDropMesoRequestHandler.kt
package net.maple.handlers.user

import client.Client
import field.obj.drop.MesoDrop
import net.maple.handlers.PacketHandler
import util.packet.PacketReader

class UserDropMesoRequestHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        val chr = c.character

        reader.readInteger() // timestamp
        val meso = reader.readInteger()

        if (meso <= chr.meso && meso > 9 && meso < 50001) {
            chr.gainMeso(-meso)
            val drop = MesoDrop(chr.id, chr, meso, 0)
            drop.field = chr.field
            drop.position = chr.position
            chr.field.enter(drop)
        } else {
            c.close(this, "Invalid meso drop amount ($meso)")
        }
    }
}